scala-api/src/test/scala/com/cloudera/livy/scalaapi/ScalaClientTest.scala

```diff
@@ -0,0 +1,245 @@
+      case Success(t) => {
+        hasTestPassed = false
+        testFailureMessage = "Test should have thrown CustomFailureException"
+        ScalaClientTest.notify(lock)
+      }
+      case Failure(e) => {
+        if (!e.getMessage.contains("CustomTestFailureException")) {
+          hasTestPassed = false
+          testFailureMessage = "Test did not throw expected exception - CustomFailureException"
+        }
+        ScalaClientTest.notify(lock)
+      }
+    }
+    ScalaClientTest.wait(lock)
+    if (!hasTestPassed) fail(testFailureMessage)
+  }
+
+  test("test Sync Rpc") {
+    configureClient(true)
+    val future = client.run(ScalaClientTest.helloJob)
+    val result = Await.result(future, 10 second)
+    assert(result === "hello")
+  }
+
+  test("test Remote client") {
+    configureClient(false)
+    val sFuture = client.submit(ScalaClientTest.simpleSparkJob)
+    val result = Await.result(sFuture, 10 second)
+    assert(result === 5)
+  }
+
+  test("test add file") {
+    configureClient(true)
+    var file: File = null
+    file = File.createTempFile("test", ".file")
+    val fileStream = new FileOutputStream(file)
+    fileStream.write("test file".getBytes("UTF-8"))
+    fileStream.close
+    val addFileFuture = client.addFile(new URI("file:" + file.getAbsolutePath()))
+    Await.ready(addFileFuture, 5 second)
+    val sFuture = client.submit(
+      context => ScalaClientTest.fileOperation(false, file.getName, context)
+    )
+    val output = Await.result(sFuture, 10 second)
+    assert(output === "test file")
+  }
+
+  test("test add jar") {
+    configureClient(true)
+    var jar: File = null
+    jar = File.createTempFile("test", ".resource")
+    val jarFile = new JarOutputStream(new FileOutputStream(jar))
+    jarFile.putNextEntry(new ZipEntry("test.resource"))
+    jarFile.write("test resource".getBytes("UTF-8"))
+    jarFile.closeEntry()
+    jarFile.close()
+    val addJarFuture = client.addJar(new URI("file:" + jar.getAbsolutePath()))
+    Await.ready(addJarFuture, 5 second)
+    val sFuture = client.submit(
+      context => ScalaClientTest.fileOperation(true, "test.resource", context)
+    )
+    val output = Await.result(sFuture, 10 second)
+    assert(output === "test resource")
+  }
+
+  test("Successive onComplete callbacks") {
+    var hasTestPassed = true
+    var testFailureMessage = ""
+    configureClient(true)
+    val future = client.run(ScalaClientTest.helloJob)
+    Await.ready(future, 5 second)
+    for (i <- 0 to 2) {
+      future onComplete {
+        case Success(t) => {
+          if (!t.equals("hello")) {
+            hasTestPassed = false;
+            testFailureMessage = "onComplete has not returned the expected message"
+          }
+        }
+        case Failure(e) => {
+          hasTestPassed = false;
+          testFailureMessage = "onComplete should not have triggered Failure callback"
+        }
+      }
+    }
+    if (!hasTestPassed) fail(testFailureMessage)
```

Comment:
This is not thread-safe. Because you're using the global execution context, callbacks may not have run when you get to this point. So if the test fails, you might miss that and report a false position.

Also, you should keep a count of how many callbacks you expect to run, since that's what you're testing here. This is another place where a semaphore would be useful.